### PR TITLE
ASoC: pcm: call snd_soc_dapm_stream_stop() in hw_free

### PR DESCRIPTION
**Pull request by ranj063**
Currently, the SND_SOC_DAPM_STREAM_START event is invoked in
pcm_prepare() but the SND_SOC_DAPM_STREAM_STOP event is
called only in dpcm_fe_dai_shutdown() after soc_pcm_close().
This results in an imbalance between when the DAPM widgets
are powered on and off.

So call snd_soc_dapm_stream_stop() in pcm_hw_clean() before
the snd_soc_pcm_component_hw_free() to keep the DAPM event
handling balanced.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
To give an insight into the sequence of the widget power up/down events: Here's the log before the change. Notice that the 
64, 1 and 2 events correspoding to WILL_PMU, PRE_PMU and POST_PMU happen after pcm hw_params. But but the events 128, 4and 8 forWILL_PMD, PRE_PMD and POST_PMD only happen after pcm_close instead of before hw_free.

```
[   28.310479] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   28.313898] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   28.313902] sof-audio-pci 0000:00:1f.3: pcm: stream dir 1, posn mailbox offset is 791060
[   28.314035] sof-audio-pci 0000:00:1f.3: received event 64 for widget DETECT9.0 --->
[   28.314038] sof-audio-pci 0000:00:1f.3: received event 1 for widget DETECT9.0 --->
[   28.314052] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   28.314269] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   28.314272] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   28.314384] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   28.314386] sof-audio-pci 0000:00:1f.3: received event 2 for widget DETECT9.0 --->
[   28.315182] sof-audio-pci 0000:00:1f.3: pcm: trigger stream 8 dir 1 cmd 1
[   28.315417] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x2024001e successful
[   28.315423] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   28.315638] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   28.315757] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   28.315778] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   28.815980] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   28.816003] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   29.256175] sof-audio-pci 0000:00:1f.3: pcm: trigger stream 8 dir 1 cmd 0
[   29.256198] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   29.256348] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   29.256355] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x240000 successful
[   29.256371] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   29.258148] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   29.258217] sof-audio-pci 0000:00:1f.3: pcm: free stream 8 dir 1
[   29.258269] sof-audio-pci 0000:00:1f.3: pcm: close stream 8 dir 1
[   29.258298] sof-audio-pci 0000:00:1f.3: received event 128 for widget DETECT9.0 --->
[   29.258302] sof-audio-pci 0000:00:1f.3: received event 4 for widget DETECT9.0 --->
[   29.258304] sof-audio-pci 0000:00:1f.3: received event 8 for widget DETECT9.0 --->
```

**Comment by ranj063** [comment]
But here's the log after the change in the PR:
```
[   62.980988] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   62.984379] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   62.984382] sof-audio-pci 0000:00:1f.3: pcm: stream dir 1, posn mailbox offset is 791060
[   62.984415] Ranjani soc_pcm_prepare
[   62.984464] sof-audio-pci 0000:00:1f.3: pcm: prepare stream 8 dir 1
[   62.984464] Ranjani soc_pcm_prepare
[   62.984482] sof-audio-pci 0000:00:1f.3: received event 64 for widget DETECT9.0 --->
[   62.984484] sof-audio-pci 0000:00:1f.3: received event 1 for widget DETECT9.0 --->
[   62.984490] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   62.984742] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60010000: GLB_STREAM_MSG: PCM_PARAMS
[   62.984744] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   62.984896] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   62.984898] sof-audio-pci 0000:00:1f.3: received event 2 for widget DETECT9.0 --->
[   62.984901] Ranjani dpcm_fe_dai_prepare
[   62.988944] sof-audio-pci 0000:00:1f.3: pcm: trigger stream 8 dir 1 cmd 1
[   62.989455] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x2024001e successful
[   62.989457] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   62.989739] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
[   62.991617] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   62.991637] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   63.491767] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   63.491795] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   63.991792] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[   63.991824] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   64.028352] sof-audio-pci 0000:00:1f.3: pcm: trigger stream 8 dir 1 cmd 0
[   64.028394] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   64.028648] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   64.028666] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x240000 successful
[   64.028676] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   64.030490] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   64.030578] sof-audio-pci 0000:00:1f.3: received event 128 for widget DETECT9.0 --->
[   64.030583] sof-audio-pci 0000:00:1f.3: received event 4 for widget DETECT9.0 --->
[   64.030586] sof-audio-pci 0000:00:1f.3: received event 8 for widget DETECT9.0 ---->
[   64.030592] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   64.030779] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   64.030786] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   64.032576] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60030000: GLB_STREAM_MSG: PCM_FREE
 ```

File: sound/soc/soc-pcm.c [hunk 1]
```diff
@@ -882,6 +882,9 @@ static int soc_pcm_hw_clean(struct snd_pcm_substream *substream, int rollback)
 			snd_soc_dai_digital_mute(dai, 1, substream->stream);
 	}
 
+	/* run the stream event */
+	snd_soc_dapm_stream_stop(rtd, substream->stream);
+
```
Comment (plbossart) [inline sound/soc/soc-pcm.c:885]: I don't get why you call this here? I would have done it after the set of hw_free() below. if you turn the power off before actually de-preparing the resources, who know what might happen.

Would moving this to line 901 make sense?

Comment (ranj063) [reply sound/soc/soc-pcm.c:885]: @plbossart the startup sequence is hw_params -> DAPM_SREAM_START. So the opposite should be DAPM_STREAM_STOP and then hw_free no? 

This change is particularly important for my other PR https://github.com/thesofproject/linux/pull/2371.

I set up the pipelines/widgets in hw_params and free them in hw_free. So ideally all DAPM power events should occur between the two to keep them balanced.

Comment (plbossart) [reply sound/soc/soc-pcm.c:885]: You may be right. I always thought hw_free was the dual of prepare, but the ALSA documentation says it's the dual of hw_params. Maybe it's the SoundWire logic that twists my brain, or what's left of it.

Comment (ranj063) [reply sound/soc/soc-pcm.c:885]: If you are good with this change, can I post this to alsa-devel @plbossart ?

Comment (plbossart) [reply sound/soc/soc-pcm.c:885]: let's check with @lgirdwood this piece of code is 8 years old 

````
01d7584cd2e5a (Liam Girdwood            2012-04-25 12:12:49 +0100 1886)         /* run the stream event for each BE */
````

Liam, any reason why the DAPM events were handled in close instead of hw_free?

Comment (plbossart) [reply sound/soc/soc-pcm.c:885]: @lgirdwood ping?

Comment (lgirdwood) [reply sound/soc/soc-pcm.c:885]: @plbossart probably related to support for OMAP4 DSP that never materialsed. I would say it's safe to move.

Comment (plbossart) [reply sound/soc/soc-pcm.c:885]: Thanks @lgirdwood can you approve so that we have your Reviewed-by tag automagically?

**Review by plbossart — APPROVED** [review]

File: sound/soc/soc-pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/soc-pcm.c:885]: @lgirdwood ping

**Comment by plbossart** [comment]
@kv2019i @lyakh @dbaluta @bardliao can we get your review?

**Review by kv2019i — DISMISSED** [review]

File: sound/soc/soc-pcm.c [hunk 2]
```diff
@@ -1858,9 +1861,6 @@ static int dpcm_fe_dai_shutdown(struct snd_pcm_substream *substream)
 	/* now shutdown the frontend */
 	soc_pcm_close(substream);
 
-	/* run the stream event for each BE */
-	dpcm_dapm_stream_event(fe, stream, SND_SOC_DAPM_STREAM_STOP);
-
```
Comment (bardliao) [inline sound/soc/soc-pcm.c:1863]: Where does ``dpcm_dapm_stream_event`` be called now?

File: sound/soc/soc-pcm.c (hunk 1 above)
Comment (bardliao) [inline sound/soc/soc-pcm.c:885]: Won't this be called twice? The other one is 
```
soc_pcm_close
	-> soc_pcm_clean(substream, 0);
		-> snd_soc_dapm_stream_stop
```

**Review by bardliao — CHANGES_REQUESTED** [review]
@ranj063 Sorry, I didn't follow the discussion, but it looks weird to add snd_soc_dapm_stream_stop, but remove dpcm_dapm_stream_event.

File: sound/soc/soc-pcm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/soc-pcm.c:885]: @bardliao you are right. This change is in the right directio but needed additional cleanups. I've fixed those now.

**Review by bardliao — APPROVED** [review]
Thanks @ranj063 Looks good now.

**Comment by plbossart** [comment]
> Thanks @ranj063 Looks good now.

great feedback @bardliao, much appreciated.

**Review by plbossart — APPROVED** [review]
looks good but I don't know enough to DPCM to really provide relevant feedack.

**Review by kv2019i — COMMENTED** [review]
I'll go ahead and merge. I find the commit message a bit hard to follow considering this patch also removes the STREAM_START (but title is just about stop), but I can't really come up with a good alternative. Current pairing does not seem right though, and hw_clean seems to be best option to send STREAM_STOP.

I recommend sending out to get feedback from other DPCM users.
